lib/trees/tree.nit

```diff
@@ -0,0 +1,614 @@
+	# ~~~
+	# str = """
+	# 1:root
+	#  2:b
+	#   3:c
+	#  4:d
+	# """
+	# tree = new Tree[Int, String].from_string(str)
+	# assert tree.length == 4
+	# assert tree[4] == "d"
+	# ~~~
+	#
+	# TODO accept more format (like with indents)
+	# FIXME This is unsafe:
+	# * returned nodes are of the type TreeNode[Int, String]
+	# * self is typed by Tree[K, E]
+	#
+	# Should we declare this as a top_level function?
+	init from_string(str: String) do
+		var lines = str.split("\n")
+		if lines.is_empty then return
+		var root = parse_line(0, lines.first)
+		add_node(root)
+		var parents = new List[N]
+		parents.add root
+		for i in [1..lines.length[ do
+			var line = lines[i]
+			if line.is_empty then continue
+			var node = parse_line(i, lines[i])
+			var space = line.space_before
+			if space > parents.last.depth then
+				add_node(node, parents.last)
+				parents.add node
+			else
+				while space <= parents.last.depth do
+					parents.pop
+				end
+				add_node(node, parents.last)
+				parents.add node
+			end
+		end
+	end
+
+	# Build this tree from a source file formatted as expected by `from_string`
+	init from_file(file: String) do
+		assert file.file_exists
+		var r = new FileReader.open(file)
+		from_string r.read_all
+		r.close
+	end
+
+	# Extract the node representation from a formated `line`
+	#
+	# Format is of the form `key:value` where `key:` is optional.
+	# If `key:` is found in string, the `key` parameter is overriden.
+	#
+	# Used by `from_string` constructor.
+	#
+	#     intrude import tree
+	#     var tree = new Tree[Int, String]
+	#     assert tree.parse_line(0, "foo").key == 0
+	#     assert tree.parse_line(0, "1:foo").key == 1
+	private fun parse_line(id: Int, line: String): N do
+		var ls = line.trim.split(":")
+		if ls.length > 1 then
+			return new TreeNode[Int, String](ls.first.to_i, ls.last)
+		end
+		return new TreeNode[Int, String](id, line.trim)
+	end
+
+	# Add a node to the tree
+	#
+	# If `parent == null` then node is set as the root node.
+	# Set `node.parent = parent` otherelse.
+	#
+	# It is recommanded to use this method instead setting the parent by hand
+	# because it also handles the `Tree::ids` and `node.children` saving.
+	fun add_node(node: N, parent: nullable N) do
+		nodes[node.key] = node
+		if parent == null then
+			root = node
+		else
+			node.parent = parent
+			parent.children.add node
+		end
+	end
+
+	# Tree nodes mapped to their key ids
+	# ~~~
+	# var str = """
+	# a
+	#  b
+	#   c
+	# """
+	# var tree = new Tree[Int, String].from_string(str)
+	# assert tree.nodes[0].value == "a"
+	# assert tree.nodes[1].value == "b"
+	# assert tree.nodes[2].value == "c"
+	# ~~~
+	#
+	# Ids can be specified in the format:
+	# ~~~
+	# str = """
+	# 10:a
+	#  20:b
+	#   0:c
+	# """
+	# tree = new Tree[Int, String].from_string(str)
+	# assert tree.nodes[10].value == "a"
+	# assert tree.nodes[20].value == "b"
+	# assert tree.nodes[0].value == "c"
+	# ~~~
+	var nodes = new HashMap[K, N]
```

Comment:
This seems to be a useful service. Maybe just type it with `Map` to disable statically write services?